crates/router/src/connector/rapyd.rs

```diff
@@ -642,25 +693,145 @@ impl services::ConnectorIntegration<api::RSync, types::RefundsData, types::Refun
+        let key = format!("whsec_verification_{}_{}", self.id(), merchant_id);
+        let secret = db
+            .get_key(&key)
+            .await
+            .change_context(errors::ConnectorError::WebhookVerificationSecretNotFound)?;
+
+        Ok(secret)
+    }
+
+    fn get_webhook_source_verification_message(
+        &self,
+        headers: &actix_web::http::header::HeaderMap,
+        body: &[u8],
+        merchant_id: &str,
+        secret: &[u8],
+    ) -> CustomResult<Vec<u8>, errors::ConnectorError> {
+        let host = conn_utils::get_header_key_value("host", headers)?;
+        let connector = self.id();
+        let url_path = format!("https://{host}/webhooks/{merchant_id}/{connector}");
+        let salt = conn_utils::get_header_key_value("salt", headers)?;
+        let timestamp = conn_utils::get_header_key_value("timestamp", headers)?;
+        let stringify_auth = String::from_utf8(secret.to_vec())
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)
+            .attach_printable("Could not convert secret to UTF-8")?;
+        let auth: transformers::RapydAuthType = stringify_auth
+            .parse_struct("RapydAuthType")
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+        let access_key = auth.access_key;
+        let secret_key = auth.secret_key;
+        let body_string = String::from_utf8(body.to_vec())
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)
+            .attach_printable("Could not convert body to UTF-8")?;
+        let to_sign = format!("{url_path}{salt}{timestamp}{access_key}{secret_key}{body_string}");
+
+        Ok(to_sign.into_bytes())
+    }
+
+    async fn verify_webhook_source(
+        &self,
+        db: &dyn StorageInterface,
+        headers: &actix_web::http::header::HeaderMap,
+        body: &[u8],
+        merchant_id: &str,
+    ) -> CustomResult<bool, errors::ConnectorError> {
+        let signature = self
+            .get_webhook_source_verification_signature(headers, body)
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+        let secret = self
+            .get_webhook_source_verification_merchant_secret(db, merchant_id)
+            .await
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+        let message = self
+            .get_webhook_source_verification_message(headers, body, merchant_id, &secret)
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+
+        let stringify_auth = String::from_utf8(secret.to_vec())
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)
+            .attach_printable("Could not convert secret to UTF-8")?;
+        let auth: transformers::RapydAuthType = stringify_auth
```

Comment:
We are parsing this twice, we can have some kind of state pattern so that we can eliminate, we can take this in different PR also.

Reply:
will take it up in differnet PR